File: app/src/main/java/org/alkaline/taskbrain/dsl/builtins/DateFunctions.kt
package org.alkaline.taskbrain.dsl.builtins

import org.alkaline.taskbrain.dsl.runtime.BuiltinFunction
import org.alkaline.taskbrain.dsl.runtime.BuiltinRegistry
import org.alkaline.taskbrain.dsl.runtime.DateTimeVal
import org.alkaline.taskbrain.dsl.runtime.DateVal
import org.alkaline.taskbrain.dsl.runtime.ExecutionException
import org.alkaline.taskbrain.dsl.runtime.StringVal
import org.alkaline.taskbrain.dsl.runtime.TimeVal
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.format.DateTimeParseException

/**
 * Date and time builtin functions.
 *
 * Milestone 2: date, datetime, time
 */
object DateFunctions {

    fun register(registry: BuiltinRegistry) {
        registry.register(dateFunction)
        registry.register(datetimeFunction)
        registry.register(timeFunction)
        registry.register(parseDateFunction)
    }

    /**
     * date - Returns the current date (no time component).
     * Dynamic: returns a different value each day.
     * Example: [date] -> 2026-01-25
     */
    private val dateFunction = BuiltinFunction(
        name = "date",
        isDynamic = true
    ) { args, _ ->
        args.requireNoArgs("date")
        DateVal(LocalDate.now())
    }

    /**
     * datetime - Returns the current date and time.
     * Dynamic: returns a different value each call.
     * Example: [datetime] -> 2026-01-25, 14:30:00
     */
    private val datetimeFunction = BuiltinFunction(
        name = "datetime",
        isDynamic = true
    ) { args, _ ->
        args.requireNoArgs("datetime")
        DateTimeVal(LocalDateTime.now())
    }

    /**
     * time - Returns the current time (no date component).
     * Dynamic: returns a different value each call.
     * Example: [time] -> 14:30:00
     */
    private val timeFunction = BuiltinFunction(
        name = "time",
        isDynamic = true
    ) { args, _ ->
        args.requireNoArgs("time")
        TimeVal(LocalTime.now())
    }

    /**
     * parse_date - Parses a string into a date value.
     * Not dynamic: pure function, same input always produces same output.
     * Example: [parse_date "2026-01-15"] -> DateVal(2026-01-15)
     *
     * Milestone 8.
     */
    private val parseDateFunction = BuiltinFunction(
        name = "parse_date",
        isDynamic = false
    ) { args, _ ->
        args.requireExactCount(1, "parse_date")
        val str = args[0] as? StringVal
            ?: throw ExecutionException("'parse_date' argument must be a string, got ${args[0]?.typeName}")
        try {
            DateVal(LocalDate.parse(str.value))
        } catch (e: DateTimeParseException) {
            throw ExecutionException("'parse_date' failed to parse date: '${str.value}'")
        }
    }
}

File: web/src/dsl/builtins/DateFunctions.ts
import { dateVal, dateTimeVal, timeVal } from '../runtime/DslValue'
import type { BuiltinFunction } from '../runtime/BuiltinRegistry'
import { ExecutionException } from '../runtime/ExecutionException'

function pad2(n: number): string {
  return String(n).padStart(2, '0')
}

function currentDate(env: { getMockedTime(): Date | undefined }): string {
  const d = env.getMockedTime() ?? new Date()
  return `${d.getFullYear()}-${pad2(d.getMonth() + 1)}-${pad2(d.getDate())}`
}

function currentTime(env: { getMockedTime(): Date | undefined }): string {
  const d = env.getMockedTime() ?? new Date()
  return `${pad2(d.getHours())}:${pad2(d.getMinutes())}:${pad2(d.getSeconds())}`
}

function currentDateTime(env: { getMockedTime(): Date | undefined }): string {
  const d = env.getMockedTime() ?? new Date()
  return `${d.getFullYear()}-${pad2(d.getMonth() + 1)}-${pad2(d.getDate())}T${pad2(d.getHours())}:${pad2(d.getMinutes())}:${pad2(d.getSeconds())}`
}

export function getDateFunctions(): BuiltinFunction[] {
  return [
    { name: 'date', isDynamic: true, call: (args, env) => {
      args.requireNoArgs('date')
      return dateVal(currentDate(env))
    }},
    { name: 'datetime', isDynamic: true, call: (args, env) => {
      args.requireNoArgs('datetime')
      return dateTimeVal(currentDateTime(env))
    }},
    { name: 'time', isDynamic: true, call: (args, env) => {
      args.requireNoArgs('time')
      return timeVal(currentTime(env))
    }},
    { name: 'parse_date', call: (args) => {
      args.requireExactCount(1, 'parse_date')
      const str = args.getPositional(0)
      if (!str || str.kind !== 'StringVal') {
        throw new ExecutionException(`'parse_date' argument must be a string, got ${str?.kind ?? 'nothing'}`)
      }
      if (!/^\d{4}-\d{2}-\d{2}$/.test(str.value)) {
        throw new ExecutionException(`'parse_date' failed to parse date: '${str.value}'`)
      }
      return dateVal(str.value)
    }},
  ]
}
